### PR TITLE
Pass time step and time step unit into QgsTemporalNavigationObject

### PR DESCRIPTION
**Pull request by Samweli**
Supersedes https://github.com/qgis/QGIS/pull/37951

Returned setFrameDuration and frameDuration functions in the QgsTemporalNavigationObject, now each time the time step and time step unit are updated in the QgsTemporalNavigationObject it's frameDuration is also updated.

### CODE REVIEW

File: src/core/qgsinterval.h [hunk 1]
```diff
@@ -187,6 +187,35 @@ class CORE_EXPORT QgsInterval
      */
     void setValid( bool valid ) { mValid = valid; }
 
+    /**
+     * Returns the interval duration.
+     *
+     * The interval duration is set only through the
```
Comment (nyalldawson) [inline src/core/qgsinterval.h:193]: I don't think this is so nice -- the other constructors could still set the unit and duration using seconds, or in the case of the y/m/d/h/min/sec constructor we could check to see if only one of the units is non zero and then use that as the duration/unit (falling back to seconds, and appropriately documenting this)

I'd also suggest we rename these to something like "originalDuration" and "originalUnit", and expand on the dox to explain how they differ from the other existing methods.

Comment (Samweli) [reply src/core/qgsinterval.h:193]: ok,

> constructor we could check to see if only one of the units is non zero and then use that as the duration/unit

I haven't quite get you here.

 I'm thinking of adding the following changes into the constructors, is this ok?

```cpp
QgsInterval::QgsInterval( double seconds )
  : mSeconds( seconds )
  , mValid( true )
{
   mOriginalDuration = mSeconds;
   mOriginalUnit = QgsUnitTypes::TemporalSeconds;
}

QgsInterval::QgsInterval( double years, double months, double weeks, double days, double hours, double minutes, double seconds )
  : mSeconds( years * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalYears, QgsUnitTypes::TemporalSeconds )
              + months * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalMonths, QgsUnitTypes::TemporalSeconds )
              + weeks * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalWeeks, QgsUnitTypes::TemporalSeconds )
              + days * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalDays, QgsUnitTypes::TemporalSeconds )
              + hours * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalHours, QgsUnitTypes::TemporalSeconds )
              + minutes * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalMinutes, QgsUnitTypes::TemporalSeconds )
              + seconds )
  , mValid( true )
{
    mOriginalDuration = mSeconds;
    mOriginalUnit = QgsUnitTypes::TemporalSeconds;
}
```

Comment (nyalldawson) [reply src/core/qgsinterval.h:193]: What I mean is something like this:

```
QgsInterval::QgsInterval( double years, double months, double weeks, double days, double hours, double minutes, double seconds )
  : mSeconds( years * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalYears, QgsUnitTypes::TemporalSeconds )
              + months * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalMonths, QgsUnitTypes::TemporalSeconds )
              + weeks * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalWeeks, QgsUnitTypes::TemporalSeconds )
              + days * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalDays, QgsUnitTypes::TemporalSeconds )
              + hours * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalHours, QgsUnitTypes::TemporalSeconds )
              + minutes * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalMinutes, QgsUnitTypes::TemporalSeconds )
              + seconds )
  , mValid( true )
{
   if ( years && !months  && !weeks  && !days && !hours && !minutes && !seconds  )
   {
      mOriginalDuration =years;
      mOriginalUnit = QgsUnitTypes::TemporalYears;
   }
   else if ( !years && months && !weeks  && !days && !hours && !minutes && !seconds  ) 
    {
      mOriginalDuration =months;
      mOriginalUnit = QgsUnitTypes::TemporalMonths;
      }
   else if ( !years && !months && weeks  && !days && !hours && !minutes && !seconds  ) 
    {
      mOriginalDuration =weeks;
      mOriginalUnit = QgsUnitTypes::TemporalWeeks;
     }
     .. .etc ...
    else
    {
       mOriginalUnit = QgsUnitTypes::TemporalUnknownUnit;
    }

}

```

Comment (Samweli) [reply src/core/qgsinterval.h:193]: ok I see, what about the constructor which takes only seconds should I also change it the way I have suggested?

Comment (nyalldawson) [reply src/core/qgsinterval.h:193]: Yep that looks good!

**Comment by Samweli** [comment]
ping @nyalldawson

File: src/core/qgsinterval.h [hunk 2]
```diff
@@ -187,6 +187,39 @@ class CORE_EXPORT QgsInterval
...
+    /**
+     * Returns the original interval duration.
+     *
+     * The interval duration is set only through the
+     * QgsInterval constructors.
+     *
+     * This original interval duration doesn't change once set,
+     * it also does not reflect the current interval duration as
+     * QgsInterval duration can change through the available
+     * QgsInterval setter methods.
```
Comment (nyalldawson) [inline src/core/qgsinterval.h:199]: Ok - a few more changes to make here:

1. Given that the "setMinutes", "setSeconds" etc methods all completely reset the interval to be a single time unit (i.e. calling setMonths with 3 months on an existing 5 day duration resets the interval to 3 months, it doesn't add to the existing 5 days to become 3months 5 days), there's no reason why we can't update the original duration and original units whenever these setters are called. That would allow us to remove this confusing behavior where the original duration may not have any resemblance at all to the interval size. (For bonus points, the various "setMinutes"/setHours setters should really explicitly mention that the existing interval size is replaced when the setter is called!) (also need tests for this)

2. The documentation for originalDuration() should mention something like "See originalUnit() for the corresponding unit. If the original duration is not available (e.g. the interval was set using a mix of units) then originalUnit() will return QgsUnitTypes::TemporalUnknownUnit."

3. The getters for months/years/days/ etc should return the original duration IF the original unit matches the requested unit. I.e. if I create a interval of 3 months duration, then calling months() on the interval should return 3.0, not the value calculated from the stored seconds value using the 30 day month assumption. (also need tests for this).  And again, the documentation for each getter should clearly state this. e.g. "Returns the interval duration in months. If the originalUnit() is QgsUnitTypes::TemporalMonths then this value will match the exact number of months as returned by originalDuration(), otherwise it will be calculated using the assumption that a month consists of exactly 30 days."

**Comment by nyalldawson** [comment]
@samweli if you can make the above requested changes then I will cherry-pick out the QgsInterval related changes from this branch and merge them alone. Then I can come back and review the rest of this work without having to split my mind between the two different parts addressed here :rofl:

**Comment by Samweli** [comment]
> @Samweli if you can make the above requested changes then I will cherry-pick out the QgsInterval related changes from this branch and merge them alone. Then I can come back and review the rest of this work without having to split my mind between the two different parts addressed here 🤣

Alright!, will do 🤣.

**Comment by Samweli** [comment]
ping @nyall

**Comment by nyalldawson** [comment]
Looks great! I'll open a PR shortly with just the QgsInterval change cherry-picked from here (plus a minor follow up improvement to the equality operator)

**Comment by nyalldawson** [comment]
https://github.com/qgis/QGIS/pull/40654

**Comment by Samweli** [comment]
ping @nyalldawson

File: src/core/qgstemporalutils.cpp [hunk 3]
```diff
@@ -141,3 +142,58 @@ bool QgsTemporalUtils::exportAnimation( const QgsMapSettings &mapSettings, const
 
   return true;
 }
+
+
+QDateTime QgsTemporalUtils::calculateFrameTime( const QDateTime &start, const long long frame, const double timeStep, const QgsUnitTypes::TemporalUnit timeStepUnit )
+{
+
+  double unused;
+  bool isFractional = fabs( modf( timeStep, &unused ) ) > 0.00001;
```
Comment (nyalldawson) [inline src/core/qgstemporalutils.cpp:151]: I think we could simplify this approach -- make calculateFrameTime accept a QgsInterval value instead of the separate timeStep/timeStepUnit values, and then retrieve the interval original unit and use that here (falling back on the fractional approach if the original interval unit is unknown unit) . That should also allow us to remove the changes to QgsTemporalNavigationObject API, since we could just keep using the QgsInterval mFrameDuration instead of introducing the new mFrameTimeStep/mFrameTimeStepUnit members.

Comment (nyalldawson) [inline src/core/qgstemporalutils.cpp:151]:
```suggestion
  const bool isFractional = !qgsDoubleNear( fabs( modf( timeStep, &unused ) ), 0.0 );
```
?

File: src/core/qgstemporalutils.h [hunk 4]
```diff
@@ -89,6 +92,25 @@ class CORE_EXPORT QgsTemporalUtils
...
+     * days, but a time step of 1 month will result in a duration that depends upon
+     * the number of days in the current month.
+     *
+     * \param start the start time of the animation.
+     * \param frame the frame number
+     * \param timeStep the time step for the animation
+     * \param timeStepUnit the time step unit for the animation
+     *
+     * \returns The calculated datetime for the frame.
+     */
```
Comment (nyalldawson) [inline src/core/qgstemporalutils.h:112]: \since QGIS 3.18

File: src/core/qgstemporalutils.h [hunk 5]
```diff
@@ -89,6 +92,25 @@ class CORE_EXPORT QgsTemporalUtils
...
+    static QDateTime calculateFrameTime( const QDateTime &start, const long long frame, double timeStep, QgsUnitTypes::TemporalUnit timeStepUnit );
```
Comment (nyalldawson) [inline src/core/qgstemporalutils.h:113]: This needs some unit tests

**Comment by rduivenvoorde** [comment]
Wow, this is a lot of changes :-) Without diving in those, I'm wondering if the original issue of: https://github.com/qgis/QGIS/pull/37951 is now actually part of all this.

I think @dminor 's issue was the fact that we used steps of 30 days in case of months, while the intuitive way is to step through the 'first' of all months (I think). Though I see the/a problem when using this in case of the 31st of a month etc (several months do not have that off course....). There are several options to handle this, it would be good if we have some clear place (in code?) on the kind of solution that is chosen?

Anyone can enlighten me?

File: src/core/qgstemporalutils.cpp (hunk 3 above)
Comment (Samweli) [reply src/core/qgstemporalutils.cpp:151]: right @nyalldawson, shouldn't we fall back to the fractional approach when the original interval duration is a fraction and not when the original interval unit is unknown unit?

**Comment by Samweli** [comment]
Hi @rduivenvoorde 

> Without diving in those, I'm wondering if the original issue of: #37951 is now actually part of all this.
> 
> I think @dminor 's issue was the fact that we used steps of 30 days in case of months

yes it is still is. The difference now is in the implementation, instead of having timestep and timestep unit members in the QgsTemporalNavigationObject, we have updated the QgsInterval class to be able to accommodate them, so now they will be accessible through QgsInterval::originalDuration() and QgsInterval::originalUnit().

File: src/core/qgstemporalutils.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/core/qgstemporalutils.cpp:151]: @Samweli I think both situations should cause the fallback fractional approach. To clarify:

- If we have a known originalUnit (e.g months), and non fractional original originalDuration: use the new approach
- if we have a known originalUnit (e.g. months) and a fractional originalDuration: use the old approach
- if we have a unknown original unit (i.e. QgsUnitTypes::TemporalUnknownUnit): use the old approach

File: src/core/qgstemporalnavigationobject.cpp [hunk 6]
```diff
@@ -74,7 +75,8 @@ QgsExpressionContextScope *QgsTemporalNavigationObject::createExpressionContextS
   std::unique_ptr< QgsExpressionContextScope > scope = qgis::make_unique< QgsExpressionContextScope >( QStringLiteral( "temporal" ) );
   scope->setVariable( QStringLiteral( "frame_rate" ), mFramesPerSecond, true );
   scope->setVariable( QStringLiteral( "frame_number" ), mCurrentFrameNumber, true );
-  scope->setVariable( QStringLiteral( "frame_duration" ), mFrameDuration, true );
+  scope->setVariable( QStringLiteral( "frame_timestep" ), mFrameDuration.originalDuration(), true );
+  scope->setVariable( QStringLiteral( "frame_timestep_unit" ), mFrameDuration.originalUnit(), true );
```
Comment (nyalldawson) [inline src/core/qgstemporalnavigationobject.cpp:79]: Can you revert these changes? Removing frame_duration may break some existing projects.

File: src/core/qgstemporalnavigationobject.cpp [hunk 7]
```diff
@@ -90,8 +92,8 @@ QgsDateTimeRange QgsTemporalNavigationObject::dateTimeRangeForFrameNumber( long
 
   const long long nextFrame = frame + 1;
 
-  const QDateTime begin = start.addSecs( frame * mFrameDuration.seconds() );
-  const QDateTime end = start.addSecs( nextFrame * mFrameDuration.seconds() );
+  QDateTime begin = QgsTemporalUtils::calculateFrameTime( start, frame, mFrameDuration );
+  QDateTime end = QgsTemporalUtils::calculateFrameTime( start, nextFrame, mFrameDuration );
```
Comment (nyalldawson) [inline src/core/qgstemporalnavigationobject.cpp:96]:
```suggestion
  const QDateTime begin = QgsTemporalUtils::calculateFrameTime( start, frame, mFrameDuration );
  const QDateTime end = QgsTemporalUtils::calculateFrameTime( start, nextFrame, mFrameDuration );
```

File: python/core/auto_generated/qgstemporalutils.sip.in [hunk 8]
```diff
@@ -35,7 +35,9 @@ returns the maximal combined temporal extent of these layers.
     {
       QgsDateTimeRange animationRange;
 
-      QgsInterval frameDuration;
+      double frameTimeStep;
+
+      QgsUnitTypes::TemporalUnit frameTimeStepUnit;
```
Comment (nyalldawson) [inline python/core/auto_generated/qgstemporalutils.sip.in:40]: Can you revert this change too please? It's an api break, and we can safely use the same approach as we do elsewhere and rely on QgsInterval here.

File: src/app/qgsanimationexportdialog.h [hunk 9]
```diff
@@ -66,6 +66,12 @@ class APP_EXPORT QgsAnimationExportDialog: public QDialog, private Ui::QgsAnimat
     //! Returns the duration of each individual frame
     QgsInterval frameInterval() const;
 
+    //! Returns the time step of each individual frame
+    double frameTimeStep() const;
+
+    //! Returns the time step unit of each individual frame
+    QgsUnitTypes::TemporalUnit frameTimeStepUnit() const;
```
Comment (nyalldawson) [inline src/app/qgsanimationexportdialog.h:73]: See above -- we don't actually need these changes anymore, the QgsInterval frameInterval method will be sufficient

File: tests/src/core/testqgstemporalnavigationobject.cpp [hunk 10]
```diff
@@ -196,8 +196,8 @@ void TestQgsTemporalNavigationObject::frameSettings()
   QCOMPARE( temporalRangeSignal.count(), 1 );
 
   navigationObject->setFrameDuration( QgsInterval( 1, QgsUnitTypes::TemporalHours ) );
-  QCOMPARE( navigationObject->frameDuration(), QgsInterval( 1, QgsUnitTypes::TemporalHours ) );
-  QCOMPARE( temporalRangeSignal.count(), 2 );
```
Comment (nyalldawson) [inline tests/src/core/testqgstemporalnavigationobject.cpp:200]: Restore this line please

File: tests/src/core/testqgstemporalnavigationobject.cpp [hunk 11]
```diff
@@ -248,7 +248,8 @@ void TestQgsTemporalNavigationObject::expressionContext()
 
   std::unique_ptr< QgsExpressionContextScope > scope( object.createExpressionContextScope() );
   QCOMPARE( scope->variable( QStringLiteral( "frame_rate" ) ).toDouble(), 30.0 );
-  QCOMPARE( scope->variable( QStringLiteral( "frame_duration" ) ).value< QgsInterval >().seconds(), 3600.0 );
+  QCOMPARE( scope->variable( QStringLiteral( "frame_timestep" ) ).value< double >(), 1.0 );
+  QCOMPARE( scope->variable( QStringLiteral( "frame_timestep_unit" ) ).value< QgsUnitTypes::TemporalUnit >(), QgsUnitTypes::TemporalUnit::TemporalHours );
```
Comment (nyalldawson) [inline tests/src/core/testqgstemporalnavigationobject.cpp:252]: See above -- please revert this

**Review by nyalldawson — COMMENTED** [review]
Sooooo close now!!

File: src/core/qgstemporalnavigationobject.cpp (hunk 6 above)
Comment (Samweli) [reply src/core/qgstemporalnavigationobject.cpp]: sure

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Superseded by https://github.com/qgis/QGIS/pull/40858

**Comment by Samweli** [comment]
> Superseded by #40858

Thanks @nyalldawson
